pkg/rbd/rbd_util.go

```diff
@@ -205,9 +224,430 @@ func rbdStatus(pOpts *rbdVolume, userID string, credentials map[string]string) (
 	return false, output, nil
 }
 
+/*
+checkRBDSnapExists, and its counterpart checkRBDVolExists, function as checks to determine if passed
+in rbdSnapshot or rbdVolume exists on the backend.
+
+**NOTE:** These functions manipulate the rados omaps that hold information regarding
+volume names as requested by the CSI drivers. Hence, these need to be invoked only when the
+respective CSI driver generated snapshot or volume name based locks are held, as otherwise racy
+access to these omaps may end up leaving them in an inconsistent state.
+
+These functions need enough information about cluster and pool (ie, Monitors, Pool, IDs filled in)
+to operate. They further require that the RequestName element of the structure have a valid value
+to operate on and determine if the said RequestName already exists on the backend.
+
+These functions populate the snapshot or the image name, its attributes and the CSI snapshot/volume
+ID for the same when succesful.
+
+These functions also cleanup omap reservations that are stale. I.e when omap entries exist and
+backing images or snapshots are missing, or one of the omaps exist and the next is missing. This is
+because, the order of omap creation and deletion are inverse of each other, and protected by the
+request name lock, and hence any stale omaps are leftovers from incomplete transactions and are
+hence safe to garbage collect.
+*/
+func checkRBDSnapExists(rbdSnap *rbdSnapshot, credentials map[string]string) (found bool, err error) {
+	// Structure members used to determine if provided rbdSnapshot exists, are checked here, to
+	// provide an easy way to check when this function can be reused
+	if rbdSnap.RequestName == "" || rbdSnap.Monitors == "" || rbdSnap.AdminID == "" ||
+		rbdSnap.Pool == "" || rbdSnap.VolName == "" || rbdSnap.ClusterID == "" {
+		return false, errors.New("missing information in rbdSnapshot to check for existence")
+	}
+
+	key, err := getRBDKey(rbdSnap.ClusterID, rbdSnap.AdminID, credentials)
+	if err != nil {
+		return false, err
+	}
+
+	// check if request name is already part of the snaps omap
+	snapName, err := util.GetOMapValue(rbdSnap.Monitors, rbdSnap.AdminID,
+		key, rbdSnap.Pool, snapsOMap, snapsOMapNameKey+rbdSnap.RequestName)
+	if err != nil {
+		return false, nil
+	}
+
+	rbdSnap.SnapName = snapName
+
+	// TODO: use listomapvals to dump all keys instead of reading them one-by-one
```

Comment:
Nit: why would listing all keys be more efficient than retrieving a couple keys?

Reply:
My thinking is as follows,
Instead of this container making 2 rados calls across the network to fetch key 1 and then 2, if it sends a list call, it saves the entire network round trip for the second fetch call.

Internally RADOS may already have cached the values for the keys, so I am not thinking along cost amortization on disk or media access.

Reply:
Fair enough. Long term efficiency, we really should just get the entire CSI off using the CLI (and associated parsing of CLI output and error codes) and instead directly use the golang wrapper API [1]

[1] https://github.com/ceph/go-ceph